File: tools/spotify_auth/spotify_client.py
import requests

from refresh import load_tokens, refresh_access_token

# this function creates the bearer authorization header using the access token
def get_bearer_header():
    tokens = load_tokens()
    access_token = tokens.get("access_token")
    return {
        "Authorization": "Bearer " + access_token
    }

# this function retrieves the currently playing track information from the Spotify API
def get_currently_playing():
    url = "https://api.spotify.com/v1/me/player/currently-playing"
    headers = get_bearer_header()

    # sends a GET request to the Spotify API to retrieve the currently playing track information
    response = requests.get(url, headers=headers)

    # checks if the access token has expired and refreshes it if necessary
    if response.status_code == 401:
        print("Access token expired. Refreshing...")
        refresh_access_token()
        headers = get_bearer_header()
        response = requests.get(url, headers=headers)
        ## eventually handle case where refresh token is also expired or spotify returns an error
    
    # creates a dictionary with the relevant track information from the response JSON
    raw_track_data = response.json()
    track_data = {
       "track_title": raw_track_data["item"]["name"],
       "artists": [artist["name"] for artist in raw_track_data["item"]["artists"]],
       "album": raw_track_data["item"]["album"]["name"],
       "album_art_url": raw_track_data["item"]["album"]["images"][0]["url"],
       "is_playing": raw_track_data["is_playing"],
       "progress_ms": raw_track_data["progress_ms"],
       "duration_ms": raw_track_data["item"]["duration_ms"]
    }

    return track_data

if __name__ == "__main__":
    get_currently_playing()

File: src/spotify_client.cpp
#include "spotify_client.h"

#include <HTTPClient.h>
#include <ArduinoJson.h>

#include "refresh.h"

String getBearerHeader(const String& accessToken) {
    return "Bearer " + accessToken;
}

SpotifyTrack getCurrentlyPlaying(String& accessToken) {
    // create empty struct
    SpotifyTrack track;

    // create http object
    HTTPClient http;

    // begin call to spotify
    http.begin("https://api.spotify.com/v1/me/player/currently-playing");
    http.setTimeout(5000);
    http.addHeader("Authorization", getBearerHeader(accessToken));

    // get response code
    int httpResponseCode = http.GET();

    // check for no track playing code
    if (httpResponseCode == 204) {
        http.end();
        return {};
    }

    // get response string
    String payload = http.getString();

    // check for access code expiration
    if (httpResponseCode == 401){

        Serial.println("Access token expired. Refreshing...");

        http.end();

        accessToken = refreshAccessToken();

        if (accessToken == ""){
            return {};
        }

        http.begin("https://api.spotify.com/v1/me/player/currently-playing");
        http.setTimeout(5000);
        http.addHeader("Authorization", getBearerHeader(accessToken));

        httpResponseCode = http.GET();
        payload = http.getString();

    }

    // check for all other errors
    if (httpResponseCode != 200) {

        Serial.println("Current track request failed.");
        Serial.println("Status code: " + String(httpResponseCode));
        Serial.println("Response: " + payload);

        http.end();
        return {};
    }

    http.end();

    // create doc object
    JsonDocument doc;
    DeserializationError error = deserializeJson(doc, payload);

    // check for parsing error
    if (error) {
        Serial.println("JSON parsing failed");
        Serial.println(error.c_str());
        return {};
    }

    // assign struct variables to track information
    track.title = doc["item"]["name"].as<String>();
    track.artist = doc["item"]["artists"][0]["name"].as<String>();
    track.album = doc["item"]["album"]["name"].as<String>();
    track.albumArtUrl = doc["item"]["album"]["images"][0]["url"].as<String>();

    track.isPlaying = doc["is_playing"].as<bool>();
    track.progressMs = doc["progress_ms"].as<long>();
    track.durationMs = doc["item"]["duration_ms"].as<long>();

    return track;
}